README.md

```diff
@@ -0,0 +1,177 @@
+xtm_queue_notify_consumer. The less often you notify, the greater the performance,
+but the greater the latency.
+Returns 0 if queue has space. Otherwise -1 with errno set to ENOBUFS.
+
+## xtm_queue_pop_ptrs
+
+Function gets up to count elements from the queue and save them in pointer array.
+Return count of extracted messages on success. Otherwise -(count + 1) with errno
+set appropriately (as in write(2), since it implies write to internal fd).
+
+## xtm_queue_consume
+
+Function flush queue pipe, according to file descriptor passed to this function.
+Return 0 on success. Otherwise -1 with errno set appropriately (as in read(2),
+since it implies read from internal fd).
+
+Examples
+--------
+
+Full example you can see in test or perf repositories, here is only simple
+example of using library.
+
+** Allocate xtm queue with appropriate flags **
+
+```c
+unsigned flags = 0;
+flags |= XTM_QUEUE_NEED_TO_CLOSE_READFD;
+flags |= XTM_QUEUE_PRODUCER_NEED_NOTIFICATIONS;
+xtm_queue = xtm_queue_new(XTM_QUEUE_SIZE, flags);
+```
+
+** Get internal fd on consumer side and pool it in the loop **
+
+```c
+int fd = xtm_queue_consumer_fd(xtm_queue);
+fd_set readset;
+FD_ZERO(&readset);
+FD_SET(fd, &readset);
+select(fd + 1, &readset, NULL, NULL, NULL);
+```
+
+** Invoke functions in case when producer thread push functions **
+
+```c
+xtm_queue_invoke_funs_all(xtm_queue);
+```
+
+** Pop pointers to data in case when producer thread push pointers **
+
+```c
+unsigned count = xtm_queue_count(xtm_queue);
+unsigned cnt = 0;
+while (cnt < count) {
+	void *ptr_array[BATCH_COUNT_MAX];
+	unsigned popcnt = (count - cnt >= BATCH_COUNT_MAX ?
+		(unsigned)BATCH_COUNT_MAX : count - cnt);
+	int rc = xtm_queue_pop_ptrs(xtm_queue, ptr_array,
+				    popcnt);
+	for (int i = 0; i < rc; i++)
+		// do something with ptr_array[i]
+	cnt += rc;
+}
+```
+
+** Delete xtm queue, when it is no longer needed **
+
+```c
+xtm_queue_delete(xtm_queue);
+```
+
+** Get internal fd on producer side, push fun to consumer thread **
+
+```c
+int fd = xtm_queue_producer_fd(xtm_queue);
+if (xtm_queue_push_fun(xtm_queue, function, pointer) < 0) {
+try_next:
+	fd_set readset;
+	FD_ZERO(&readset);
+	FD_SET(fd, &readset);
+	// Wait fror consumer notification, that queue is not full
+	select(fd + 1, &readset, NULL, NULL, NULL);
+	xtm_queue_consume(fd);
+	if (xtm_queue_push_fun(xtm_queue, function, pointer) < 0)
+		goto try_next;
+}
+```
+
+** Get internal fd on producer side, push ptr to consumer thread **
+
+```c
+if (xtm_queue_push_ptr(xtm_queue, pointer) < 0) {
+try_next:
+	fd_set readset;
+	FD_ZERO(&readset);
+	FD_SET(fd, &readset);
+	// Wait fror consumer notification, that queue is not full
+	select(fd + 1, &readset, NULL, NULL, NULL);
+	xtm_queue_consume(fd);
+	if (xtm_queue_push_ptr(xtm_queue, pointer) < 0)
+		goto try_next;
+}
+```
+
+** Wait in the loop until queue became not full **
+
+```c
+while (xtm_msg_probe(queue) != 0);
```

Comment:
This looks like a usecase we didn't cover. If a user wants to wait for a queue being non-full, or having at least N free elements, but not push anything until that, he indeed can only wait in a busy loop which we can not allow. Even if we would make `probe()` in case of the full queue set the flag for getting a notification from the producer, it won't help in case a user wants to wait for N free slots. Maybe a new function like `xtm_queue_ask_producer_notify()` could help. Or `xtm_queue_probe_count(N)` which would ensure there is at least N slots and otherwise ask for more via the flag. But we can do it later, not a blocker.